k2/csrc/array.h

```diff
@@ -161,24 +161,33 @@ struct Array3 {
   using IndexT = I;
   using PtrT = Ptr;
 
-  IndexT size;
-  IndexT *indexes1;  // indexes1[0,1,...size] should be defined; note,
-                     // this means the array must be of at least size+1.
+  IndexT size1;  // equal to the number of Array2 object in this Array3 object;
+                 // `size1 + 1` will be the number of elements in indexes1.
+
+  IndexT size2;  // equal to indexes1[size1] - indexes1[0];
+                 // `size2 + 1` will be the number of elements in indexes2;
+
+  IndexT size3;  // the number of elements in `data`,  equal to
+                 // indexes2[indexes1[size1]] - indexes2[indexes1[0]].
+
+  IndexT *indexes1;  // indexes1[0,1,...size1] should be defined; note,
+                     // this means the array must be of at least size1+1.
                      // We require that indexes[i] <= indexes[i+1], but it
-                     // is not required that indexes[0] == 0, it may be
-                     // greater than 0.
+                     // is not required that indexes[0] == 0, it may be greater
+                     // than 0.
 
   IndexT *indexes2;  // indexes2[indexes1[0]]
-                     // .. indexes2[indexes1[size]-1] should be defined.
+                     // .. indexes2[indexes1[size1]] should be defined;
```

Comment:
note that `indexes2` has size `size2 + 1` instead of `size2` now to be consistent with the requirement of `indexes` in `Array2`